src/shell/shell_cmds.cpp

```diff
@@ -608,6 +613,52 @@ static std::vector<int> calc_column_widths(const std::vector<int> &word_widths,
 	return col_widths;
 }
 
+char buffer[15] = {0};
+char *FormatDate(uint16_t year, uint8_t month, uint8_t day)
+{
+	char formatstring[6], c = dos.tables.country[11];
+	sprintf(formatstring,
+	        dos.tables.country[0] == 1
+	                ? "D%cM%cY"
+	                : (dos.tables.country[0] == 2 ? "Y%cM%cD" : "M%cD%cY"),
+	        c, c);
+	Bitu bufferptr = 0;
+	for (Bitu i = 0; i < 5; i++) {
+		if (i == 1 || i == 3) {
+			buffer[bufferptr] = formatstring[i];
+			bufferptr++;
+		} else {
+			if (formatstring[i] == 'M')
+				bufferptr += (Bitu)sprintf(buffer + bufferptr,
+				                           "%02u", month);
+			if (formatstring[i] == 'D')
+				bufferptr += (Bitu)sprintf(buffer + bufferptr,
+				                           "%02u", day);
+			if (formatstring[i] == 'Y')
+				bufferptr += (Bitu)sprintf(buffer + bufferptr,
+				                           "%04u", year);
+		}
+	}
+	return buffer;
+}
+
+char *FormatTime(Bitu hour, Bitu min, Bitu sec, Bitu msec)
+{
+	Bitu fhour = hour;
+	static char retBuf[14];
+	char ampm[3] = "";
+	if (!(dos.tables.country[17] & 1)) { // 12 hour notation?
+		if (hour != 12)
+			hour %= 12;
+		strcpy(ampm, hour != 12 && hour == fhour ? "am" : "pm");
+	}
+	char sep = dos.tables.country[13];
+	sprintf(retBuf, "%u%c%02u%c%02u%c%02u%s", (unsigned int)hour,
+	        sep, (unsigned int)min, sep, (unsigned int)sec,
+	        dos.tables.country[9], (unsigned int)msec, ampm);
+	return retBuf;
+}
+
```

Comment:
Is there a reason for rolling your own date/time string formatting, instead of using `strftime`? For example, for date, I might have used a format string of `%Y/%m/%d`, `%d/%m/%Y` or `%m/%d/%Y`, and and replaced the separator character afterwards if required.

I'll concede these functions may be slightly more efficient.

Reply:
@shermp The updated date formatting code is based on the existing date format code, which tried to format the date using bufferptr. Perhaps this one can be further cleaned up with strfttime or so later.

Reply:
Ok, yeah, we can revisit later I guess